pilot/pkg/proxy/envoy/v1/gateway.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2018 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	"strings"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	routing "istio.io/api/routing/v1alpha2"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+)
+
+func buildGatewayListeners(mesh *meshconfig.MeshConfig,
+	instancesOnThisNode []*model.ServiceInstance,
+	allServicesInTheMesh []*model.Service,
+	config model.IstioConfigStore, node model.Node) Listeners {
+
+	gateways, err := config.List(model.Gateway.Type, model.NamespaceAll)
+	if err != nil {
+		log.Warnf("error listing gateways: %v", err)
+		return Listeners{}
+	} else if len(gateways) == 0 {
+		log.Debugf("listing gateways in namespace %q, got none", model.NamespaceAll)
+		return Listeners{}
+	}
+
+	targetGateways := map[string]bool{}
+	for _, spec := range gateways {
+		targetGateways[spec.Name] = true
+	}
+
+	// TODO: is this still right??
+	gateway := gateways[0].Spec.(*routing.Gateway)
+	for _, spec := range gateways[1:] {
+		model.MergeGateways(gateway, spec.Spec.(*routing.Gateway))
+	}
+
+	listeners := make(Listeners, 0, len(gateway.Servers))
+	for _, server := range gateway.Servers {
+		// TODO: TCP
+
+		// build physical listener
+		physicalListener := buildPhysicalGatewayListener(mesh, node, instancesOnThisNode, allServicesInTheMesh, config, server)
```

Comment:
nit. buildGatewayHTTPListener

Reply:
Oh ignore.. may be buildGatewayListener or buildSingleGatewayListener or something

Reply:
ignoring.